wolfcrypt/src/ecc.c

```diff
@@ -4239,13 +4239,27 @@ static int wc_ecc_sign_hash_hw(const byte* in, word32 inlen,
 #endif
     {
         word32 keysize = (word32)key->dp->size;
+        word32 orderBits;
+        DECLARE_CURVE_SPECS(curve, 1);
 
         /* Check args */
-        if (keysize > ECC_MAX_CRYPTO_HW_SIZE || inlen != keysize ||
-                                                *outlen < keysize*2) {
+        if (keysize > ECC_MAX_CRYPTO_HW_SIZE || *outlen < keysize*2) {
             return ECC_BAD_ARG_E;
         }
 
+        /* if the input is larger than curve order, we must truncate */
+        ALLOC_CURVE_SPECS(1);
+        err = wc_ecc_curve_load(key->dp, &curve, ECC_CURVE_FIELD_ORDER);
+        if (err != 0) {
+           FREE_CURVE_SPECS();
+           return err;
+        }
+        orderBits = mp_count_bits(curve->order);
+        if ((inlen * WOLFSSL_BIT_SIZE) > orderBits) {
+           inlen = (orderBits + WOLFSSL_BIT_SIZE - 1) / WOLFSSL_BIT_SIZE;
+        }
+        FREE_CURVE_SPECS();
```

Comment:
Do we need to worry about orderBits that aren't divisible by 8, like 239 bit or 521bit?  Or do we only support hardware that uses curves divisible by 8 bits?

Reply:
That's handled with the `(orderBits + WOLFSSL_BIT_SIZE - 1) / WOLFSSL_BIT_SIZE` logic, which is the same used below. At this time only 256 and 384 bits keys are supported with hardware, but I wanted to make sure inlen would get truncated correctly in the future.

Reply:
Not exactly, in the case of 239 this would truncate a 256bit hash to 240 bits (leaving on an extra bit).  The software version would right shift one more bit with mp_rshb() after the rounded up byte truncation.  Looks like we don't need to support that for now.